synapse/axon.py

```diff
@@ -850,3 +867,366 @@ def has(self, htype, hvalu):
+
+        return 0
+
+    def fs_getattr(self, path):
+        '''
+        Return the file attributes for a given file path.
+
+        Example:
+
+            axon.fs_getattr('/foo/bar/baz.faz')
+
+        '''
+        path = self._fs_normpath(path)
+        tufo = self.core.getTufoByProp('axon:path', path)
+        return Axon._fs_tufo2attr(tufo)
+
+    def fs_getxattr(self, path, name):
+        '''
+        Return a file attribute value for a given file path and attr name.
+
+        Example:
+
+            axon.fs_getxattr('/foo/bar/baz.faz', 'st_size')
+
+        '''
+        if name not in _fs_attrs:
+            raise NoSuchData()
+
+        path = self._fs_normpath(path)
+        tufo = self.core.getTufoByProp('axon:path', path)
+        if tufo:
+            return tufo[1].get('axon:path:%s' % name)
+
+        raise NoSuchData()
+
+    def fs_mkdir(self, path, mode):
+        '''
+        Creates a new directory at the given path.
+        Returns 0.
+
+        Example:
+
+            axon.fs_mkdir('/mydir')
+
+        '''
+        normed, props = self.core.getPropNorm('axon:path', path)
+        ppath = props.get('dir')
+
+        if not ppath:
+            raise NoSuchDir()
+
+        parentfo = self.core.getTufoByProp('axon:path', ppath)
+        if not(parentfo and Axon._fs_isdir(parentfo[1].get('axon:path:st_mode'))):
+            raise NoSuchDir()
+
+        attr = Axon._fs_new_dir_attrs(ppath, mode)
+        tufo = self.core.formTufoByProp('axon:path', path, **attr)
+        if tufo and tufo[1].get('.new') != True:
+            raise FileExists()
+
+        self.core.incTufoProp(parentfo, 'st_nlink', 1)
+
+    def fs_read(self, path, size, offset):
+        '''
+        Reads a directory.
+        Returns list of files.
+
+        Example:
+
+            axon.fs_read('/mydir', 100, 0)
+
+        '''
+        tufo = self.core.getTufoByProp('axon:path', path)
+        if not tufo:
+            raise NoSuchEntity()
+        bval = tufo[1].get('axon:path:blob')
+        blob = None
+
+        if bval:
+            blob = self.core.getTufoByProp('axon:blob', bval)
+
+        if not blob:
+            return b''
+
+        boff = blob[1].get('axon:blob:off')
+        blob[1]['axon:blob:off'] = boff + offset  # the offset of the blob in the axon + the offset within the file
+        blob[1]['axon:blob:size'] = size  # number of bytes that the OS asks for
+
+        return b''.join(self.iterblob(blob))
+
+    def fs_readdir(self, path):
+        '''
+        Reads a directory.
+        Returns list of files.
+
+        Example:
+
+            axon.fs_readdir('/mydir')
+
+        '''
+        files = ['.','..']
+
+        attr = self.fs_getattr(path)
+        if not Axon._fs_isdir(attr.get('st_mode')):
+            raise NotSupported()
+
+        tufos = self.core.getTufosByProp('axon:path:parent', path)
+        for tufo in tufos:
+            fpath = tufo[1].get('axon:path')
+            fname = fpath.split('/')
```

Comment:
nit: fname = fpath.split('/')[-1] will be faster since we can avoid the extra calls will still being pretty readable